### DISCUSSION TITLE
GPU usage of Ascent

### DISCUSSION THREAD

**Discussion by yslan** [Alpine-DAV/ascent#1267] [General]
Hi

1. When we specify `runtine/vtkm/backend` what happen if the backend is not availiable? Will it choses the CPU / OpenMP instead?

   I know we can print out the about() or the runtime actions but I'm not sure those info are indeed used since we print it outside Ascent's API.

   Or, maybe there is an API function to confirm which backend is used? 

2. What GPU stream does Ascent use? Is there a already a synchronization in Ascent's API?  

Thanks!

**Comment by nicolemarsaglia** [comment]
@yslan I did not test this, but most likely, if the specified backend is not available it will break expecting to be built with the correct flags and TPLs. Otherwise it should default to serial on the CPU. 

Yes, another way to get information from Ascent using the API is either of the info() calls. This will fill or produce a conduit node with information that you can print out: https://github.com/Alpine-DAV/ascent/blob/950c5f18e47bea2432113516f543e28136555095/src/libs/ascent/ascent_runtime.hpp#L38

https://github.com/Alpine-DAV/ascent/blob/950c5f18e47bea2432113516f543e28136555095/src/libs/ascent/ascent_runtime.hpp#L40


In terms of what GPUs Ascent will work on, it is capable of using VTK-m with CUDA, and VTK-m with Kokkos and HIP. I'm currently working on an implementation using VTK-m with Kokkos and CUDA. 

Let me know if you have any more questions!

**Reply by yslan** [reply]
@nicolemarsaglia  Thanks for the response. 
I'm actually wondering about the `cudaStream_t` that could cover kernel launch time and memcopy. 
https://developer.download.nvidia.com/CUDA/training/StreamsAndConcurrencyWebinar.pdf

**Reply by nicolemarsaglia** [reply]
@yslan ah, now I see, interesting! Thanks for correcting me, I'd never heard of GPU streams before. I'm not sure of the answer. I don't think Ascent does any GPU streaming or synchronization directly, that is handled by either VTKm or Kokkos, and even then I'm not sure if either of them have it, more likely the latter than the former. 

I'll add to this if I find out more. Or @cyrush can add his knowledge when he is back next week.

**Comment by yslan** [comment]
After digging into the code a bit further, I found vtkh's API `vtkh::AboutVTKH()`, `vtkh::IsCUDAEnabled()`, `vtkh::ForceCUDA()` in [src/libs/vtkh/vtkh.hpp](https://github.com/Alpine-DAV/ascent/blob/950c5f18e47bea2432113516f543e28136555095/src/libs/vtkh/vtkh.hpp) which could be useful to double check what backend is actually used. 

Otherwise, based on the vtkm [code](https://gitlab.kitware.com/vtk/vtk-m/-/blob/master/vtkm/cont/Initialize.cxx?ref_type=heads#L228-241), it will reset the backend if the proposed one is not available.
It is nice to keep the code running. However, I do think user should be aware if the desired backend (says cuda) is not working and the Ascent could actually fall back to OpenMP and introduce performance regression. 

`Info()` from the conduit node prints filters / expressions but not GPU backend.

**Comment by yslan** [comment]
@nicolemarsaglia and @cyrush 

Unfortunately, if I have multiple backends, vtkm will show something like this.
It's still not clear what is used.
```
---------------- VTK-h -------------------
MPI version: 3.1
VTK-m adapters: Cuda (enabled) Serial (enabled)
------------------------------------------
```

Let's assume we use vtkh/vtkm. 

1. Ascent's vtkh doesn't seem to specify the default backend?

2. What's the default backend for vtkm? 
    I got lost when I try to see the source code and their mailing list is not public available.

3. How can we confirm which backend is used without profiling the code?

I appreciate any inputs. Thanks in advance!

**Comment by cyrush** [comment]
@yslan 
If you want to force cuda, you can use:

```
Node opts;
opts["runtine/vtkm/backend"] = "cuda";
// ... other mpi options, etc
ascent.open(opts);
```

This will call VTK-h's "Force Cuda" routine, which should configure VTK-m to only allow cuda and not fall back to serial.

https://github.com/Alpine-DAV/ascent/blob/950c5f18e47bea2432113516f543e28136555095/src/libs/ascent/ascent.cpp#L273

If possible, I would still run once with a device profiler to check everything is on device.

**Reply by yslan** [reply]
> This will call VTK-h's "Force Cuda" routine

Yeah... this is something I tried to avoid because it needs some checking and there are many options.
And, so far, it appears that I need to do many profiling to make sure this is working.

For example, say we want to use kokkos-HIP. If it's not availiable, maybe fall back to OpenMP. If not, then serial. 
Such workflow should be inside vtkh's API? and user provide the list of the attempts.

> If possible, I would still run once with a device profiler to check everything is on device.

This is not ideal since we will have to check on every machines.

**Reply by cyrush** [reply]
I understand.

The default VTK-m behavior should be device, then a backup. I don't know off hand if it will choose OpenMP or Serial as backup. 

It's hard to do experiments like this b/c backup should only be triggered if there is a problem running on the device (like things are compiled wrong, maybe for the wrong gpu architecture -- which would compile but cause a problem at runtime)

That's why we have the force -- you get what you ask for or you fail.

Would it be helpful to have a setting that is "device", instead of "cuda" -- so you can force device regardless of nvidia or amd?

kokkos clouds this bit - but as of right now we are only using kokkos for hip.

I will see if we can expose what was actually used in ascent::info() - which would help with reporting. 

Would you like ascent.open() to have an option like `["device","openmp","serial"] to control precedence?

We may need to work with VTK-m folks to get better control options and feedback.

**Reply by yslan** [reply]
I want to make sure it uses the exact same backend user provide.

I can set default backend using the current vtkh's API.
Then, I think user is on their own to make sure it's runable. 

I guess I'm lucky that all machines I'm using only have single GPU vendor so we eliminate combinations by trusting Ascent experts to install it properly.

One extra question:
What's your current plan for Inter GPU? kokkos? OpenMP?

**Reply by cyrush** [reply]
@yslan the best path is to use ascent options - not the vtk-h api directly, so let's make sure whatever you need you can access it via that API. 

Argonne folks have been working on testing Intel GPU Support. VTK-m will use kokkos using SYCL, and for the places were we use RAJA that will also be the path.

**Reply by yslan** [reply]
Then, I need some "feedback" from Ascent's interface like whether the backend is available or whether it's working, such as `IsCUDAAvailable` and `IsCUDAEnable`.

**Reply by stgeke** [reply]
How about ["cuda","Kokkos::HIP","openmp","serial"]? This way you get what you ask for or you fail.

**Reply by cyrush** [reply]
sounds good, I'll work on the specific names and the expected semantics and create a short write up as an issue

**Reply by cyrush** [reply]
Here is an issue with some initial ideas to evolve:

https://github.com/Alpine-DAV/ascent/issues/1273
